core/modules/path/path.module

```diff
@@ -453,10 +808,875 @@ function path_entity_update(Entity $entity) {
+ * @param $redirect
+ *   The redirect object on which the operation is to be performed, or redirect
+ *   type (e.g. 'feedburner') for the "create" operation.
+ * @param $account
+ *   Optional, a user object representing the user for whom the operation is to
+ *   be performed. Determines access for a user other than the current user.
+ *
+ * @return
+ *   TRUE if the operation may be performed, FALSE otherwise.
+ */
+function redirect_access($op, $redirect, $account = NULL) {
+  global $user;
+
+  $rights = &backdrop_static(__FUNCTION__, array());
+
+  if (!$redirect || !in_array($op, array('create', 'update', 'delete'), TRUE)) {
+    // If there was no redirect to check against, or the $op was not one of the
+    // supported ones, we return access denied.
+    return FALSE;
+  }
+  // If no user object is supplied, the access check is for the current user.
+  if (empty($account)) {
+    $account = $user;
+  }
+
+  $cid = isset($redirect->rid) ? $redirect->rid : $redirect;
+
+  // Return cached value if access already checked for this redirect, user and op.
+  if (isset($rights[$account->uid][$cid][$op])) {
+    return $rights[$account->uid][$cid][$op];
+  }
+
+  // Administrators can access all redirects.
+  if (user_access('administer redirects', $account)) {
+    $rights[$account->uid][$cid][$op] = TRUE;
+    return TRUE;
+  }
+
+  // We grant access to the redirect if both of the following conditions are met:
+  // - No modules say to deny access.
+  // - At least one module says to grant access.
+  $access = module_invoke_all('redirect_access', $op, $redirect, $account);
+  if (in_array(REDIRECT_ACCESS_DENY, $access, TRUE)) {
+    $rights[$account->uid][$cid][$op] = FALSE;
+    return FALSE;
+  }
+  elseif (in_array(REDIRECT_ACCESS_ALLOW, $access, TRUE)) {
+    $rights[$account->uid][$cid][$op] = TRUE;
+    return TRUE;
+  }
+
+  return FALSE;
+}
+
+/**
+ * Validate a redirect.
+ */
+function redirect_validate(Redirect $redirect, $form, &$form_state) {
+  // Check that there there are no redirect loops.
+  if (url($redirect->source) == url($redirect->redirect)) {
+    form_set_error('redirect', t('You are attempting to redirect the page to itself. This will result in an infinite loop.'));
+  }
+
+  redirect_hash($redirect);
+  if ($existing = redirect_load_by_hash($redirect->hash)) {
+    if ($redirect->rid != $existing->rid) {
+      form_set_error('source', t('The source path %source is already being redirected. Do you want to <a href="@edit-page">edit the existing redirect</a>?', array('%source' => redirect_url($redirect->source, $redirect->source_options), '@edit-page' => url('admin/config/search/redirect/edit/'. $existing->rid))));
+    }
+  }
+
+  // Allow other modules to validate the redirect.
+  foreach (module_implements('redirect_validate') as $module) {
+    $function = $module . '_redirect_validate';
+    $function($redirect, $form, $form_state);
+  }
+}
+
+function redirect_object_prepare($redirect, $defaults = array()) {
```

Comment:
Needs docblock.

`$redirect` parameter should probably be hinted to `Redirect $redirect`